FreeRTOS_TCP_IP.c

```diff
@@ -3583,9 +3583,9 @@
                                 vTCPStateChange( pxSocket, eCLOSED );
                             }
                             /* Otherwise, check whether the packet is within the receive window. */
-                            else if( ( ulSequenceNumber > pxSocket->u.xTCP.xTCPWindow.rx.ulCurrentSequenceNumber ) &&
-                                     ( ulSequenceNumber < ( pxSocket->u.xTCP.xTCPWindow.rx.ulCurrentSequenceNumber +
-                                                            pxSocket->u.xTCP.xTCPWindow.xSize.ulRxWindowLength ) ) )
+                            else if( ( xSequenceGreaterThan( ulSequenceNumber, pxSocket->u.xTCP.xTCPWindow.rx.ulCurrentSequenceNumber ) ) &&
+                                     ( xSequenceLessThan( ulSequenceNumber, ( pxSocket->u.xTCP.xTCPWindow.rx.ulCurrentSequenceNumber +
+                                                                              pxSocket->u.xTCP.xTCPWindow.xSize.ulRxWindowLength ) ) ) )
```

Comment:
you could also do something like this:
```C
else if ( ulSequenceNumber - pxSocket->u.xTCP.xTCPWindow.rx.ulCurrentSequenceNumber <
          pxSocket->u.xTCP.xTCPWindow.xSize.ulRxWindowLength )
```

Reply:
Yes I believe that. indirectly, that test has been done. Formally ( i.e. without the overflows ), the following two staments are equal:

- else if ( ulSequenceNumber - ulCurrentSequenceNumber < ulRxWindowLength )
- else if ( ulSequenceNumber < ulRxWindowLength + ulCurrentSequenceNumber )

Only I have expressed it as:

 ( xSequenceLessThan( ulSequenceNumber, ( ulCurrentSequenceNumber + ulRxWindowLength ) ) )

Reply:
xSequenceLessThan is not linking.  Did a file get missed in the commit?

Reply:
right, both ways it is correct as long as `ulRxWindowLength <= 2^31`
it's just that your solution needs two checks.

Reply:
actually to prove the equivalence I need `1 <= ulRxWindowLength && ulRxWindowLength <= 2^31+1`
Are we sure that  ulRxWindowLength is not zero?

Reply:
@n9wxu 
> 
> xSequenceLessThan is not linking. Did a file get missed in the commit?

Sorry yes, I forgot to upload the TCP_WIN source files. One moment.

Reply:
> Are we sure that ulRxWindowLength is not zero?

I gave a quick look to the source and it seems that `ulRxWindowLength` is never 0. Even when the windowing is disabled for TCP, the value of `ulRxWindowLength == ipconfigTCP_MSS`.

If someone was to mess with `ipconfigTCP_MSS` and make it 0, then the overflow of sequence numbers would be the least of their problems. :)